### ISSUE TITLE
llvm-spirv error on -O0 opt level

### ISSUE THREAD

**Issue by akharche** [IntelPython/numba-dppy#252]
If I change the opt level to -O0 inside the [spirv-generator](https://github.com/IntelPython/numba-dppy/blob/main/numba_dppy/spirv_generator.py#L85), I get the following error:

To reproduce it run examples/sum.py with -O0 instead of -O1

```
llvm-spirv: /home/sat_bot/base/conda-bld/llvm-spirv_1605662563618/work/lib/SPIRV/SPIRVWriter.cpp:706: SPIRV::SPIRVInstruction* SPIRV::LLVMToSPIRV::transUnaryInst(llvm::UnaryInstruction*, SPIRV::S
PIRVBasicBlock*): Assertion `Cast->getSrcTy()->getPointerAddressSpace() == SPIRAS_Generic' failed.
Stack dump:
0.      Program arguments: llvm-spirv -o /tmp/tmpcd4h1ict/1-generated-spirv /tmp/tmpcd4h1ict/0-llvm-friendly-spir.bc
1.      Running pass 'LLVMToSPIRV' on module '/tmp/tmpcd4h1ict/0-llvm-friendly-spir.bc'.
 #0 0x00007f9c55e52cbb llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVM-10.so+0xa81cbb)
 #1 0x00007f9c55e50814 llvm::sys::RunSignalHandlers() (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVM-10.so+0xa7f814)
 #2 0x00007f9c55e50964 SignalHandler(int) (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVM-10.so+0xa7f964)
 #3 0x00007f9c5a8ad3c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x153c0)
 #4 0x00007f9c54f4d18b raise /build/glibc-ZN95T4/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #5 0x00007f9c54f2c859 abort /build/glibc-ZN95T4/glibc-2.31/stdlib/abort.c:81:7
 #6 0x00007f9c54f2c729 get_sysdep_segment_value /build/glibc-ZN95T4/glibc-2.31/intl/loadmsgcat.c:509:8
 #7 0x00007f9c54f2c729 _nl_load_domain /build/glibc-ZN95T4/glibc-2.31/intl/loadmsgcat.c:970:34                                                                                                      #8 0x00007f9c54f3df36 (/lib/x86_64-linux-gnu/libc.so.6+0x36f36)
 #9 0x00007f9c5a75142c SPIRV::LLVMToSPIRV::transUnaryInst(llvm::UnaryInstruction*, SPIRV::SPIRVBasicBlock*) (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVMSPIRVLib.so.10+0x1dc42c)
#10 0x00007f9c5a74b246 SPIRV::LLVMToSPIRV::transValueWithoutDecoration(llvm::Value*, SPIRV::SPIRVBasicBlock*, bool) (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVMSPIRVLib.
so.10+0x1d6246)
#11 0x00007f9c5a74bd22 SPIRV::LLVMToSPIRV::transValue(llvm::Value*, SPIRV::SPIRVBasicBlock*, bool) (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVMSPIRVLib.so.10+0x1d6d22)
#12 0x00007f9c5a751f1e SPIRV::LLVMToSPIRV::transFunction(llvm::Function*) (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVMSPIRVLib.so.10+0x1dcf1e)
#13 0x00007f9c5a753f11 SPIRV::LLVMToSPIRV::translate() (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVMSPIRVLib.so.10+0x1def11)
#14 0x00007f9c5a7540b8 SPIRV::LLVMToSPIRV::runOnModule(llvm::Module&) (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVMSPIRVLib.so.10+0x1df0b8)                               #15 0x00007f9c55f67041 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVM-10.so+0xb96041)
#16 0x00007f9c5a7541b3 llvm::writeSpirv(llvm::Module*, SPIRV::TranslatorOpts const&, std::ostream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (/localdisk/w
ork/akharche/miniconda3/envs/dppy_1301/bin/../lib/libLLVMSPIRVLib.so.10+0x1df1b3)
#17 0x000055d540af076f main (/localdisk/work/akharche/miniconda3/envs/dppy_1301/bin/llvm-spirv+0xa76f)
#18 0x00007f9c54f2e0b3 __libc_start_main /build/glibc-ZN95T4/glibc-2.31/csu/../csu/libc-start.c:342:3
#19 0x000055d540af0a07 _start /home/rdonnelly/mc/conda-bld/compilers_linux-64_1534865402226/work/.build/src/glibc-2.12.2/csu/../sysdeps/x86_64/elf/start.S:116:0
Traceback (most recent call last):
  File "sum.py", line 62, in <module>
    main()
  File "sum.py", line 50, in main
    driver(a, b, c, global_size)
  File "sum.py", line 35, in driver
    data_parallel_sum[global_size, dppy.DEFAULT_LOCAL_SIZE](a, b, c)
  File "/localdisk/work/akharche/numba-dppy/numba_dppy/compiler.py", line 570, in __call__
    kernel = self.specialize(*args)
  File "/localdisk/work/akharche/numba-dppy/numba_dppy/compiler.py", line 591, in specialize
    self.sycl_queue, self.py_func, argtypes, self.access_types
  File "/localdisk/work/akharche/numba-dppy/numba_dppy/compiler.py", line 165, in compile_kernel
    ordered_arg_access_types=access_types,
  File "/localdisk/work/akharche/numba-dppy/numba_dppy/compiler.py", line 394, in __init__
    self.spirv_bc = spirv_generator.llvm_to_spirv(self.context, self.binary)
  File "/localdisk/work/akharche/numba-dppy/numba_dppy/spirv_generator.py", line 198, in llvm_to_spirv
  return mod.finalize()
  File "/localdisk/work/akharche/numba-dppy/numba_dppy/spirv_generator.py", line 148, in finalize
    self._cmd.generate(ipath=self._llvmfile, opath=spirv_path)
  File "/localdisk/work/akharche/numba-dppy/numba_dppy/spirv_generator.py", line 86, in generate
    check_call(["llvm-spirv", "-o", opath, ipath + ".bc"])
  File "/localdisk/work/akharche/numba-dppy/numba_dppy/spirv_generator.py", line 40, in check_call
    return _real_check_call(*args, **kwargs)
  File "/localdisk/work/akharche/miniconda3/envs/dppy_1301/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['llvm-spirv', '-o', '/tmp/tmpcd4h1ict/1-generated-spirv', '/tmp/tmpcd4h1ict/0-llvm-friendly-spir.bc']' died with <Signals.SIGABRT: 6>.
Exception ignored in: <function Module.__del__ at 0x7fcd09803a70>
Traceback (most recent call last):
  File "/localdisk/work/akharche/numba-dppy/numba_dppy/spirv_generator.py", line 117, in __del__
    os.rmdir(self._tmpdir)
OSError: [Errno 39] Directory not empty: '/tmp/tmpcd4h1ict'

```

**Comment by akharche** [comment]
_LLVM version 10.0.0_
_Libllvm10
llvm-spirv  - 10.0.0
llvm-tools - 10.0.0 
llvmdev  - 10.0.0 
Llvmlite - 0.35.0_


Switching to 
_llvm-spirv  - 11.0.0
llvm-tools - 11.0.0_

Gives the following error

```
llvm-spirv: /home/sat_bot/base/conda-bld/llvm-spirv_1614178331572/work/lib/SPIRV/SPIRVWriter.cpp:869: SPIRV::SPIRVInstruction* SPIRV::LLVMToSPIRV::transUnaryInst(llvm::UnaryInstruction*, SPIRV::SPIRVBasic
Block*): Assertion `SrcAddrSpace == SPIRAS_Generic' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: llvm-spirv -o /tmp/tmpygxj1cph/1-generated-spirv /tmp/tmpygxj1cph/0-llvm-friendly-spir.bc
1.      Running pass 'LLVMToSPIRV' on module '/tmp/tmpygxj1cph/0-llvm-friendly-spir.bc'.
 #0 0x00007f39f8da536b llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVM-11.so+0xb0736b)
 #1 0x00007f39f8da31d4 llvm::sys::RunSignalHandlers() (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVM-11.so+0xb051d4)
 #2 0x00007f39f8da3329 SignalHandler(int) (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVM-11.so+0xb05329)
 #3 0x00007f39fe1213c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x153c0)
 #4 0x00007f39f7e1a18b raise /build/glibc-ZN95T4/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #5 0x00007f39f7df9859 abort /build/glibc-ZN95T4/glibc-2.31/stdlib/abort.c:81:7
 #6 0x00007f39f7df9729 get_sysdep_segment_value /build/glibc-ZN95T4/glibc-2.31/intl/loadmsgcat.c:509:8
 #7 0x00007f39f7df9729 _nl_load_domain /build/glibc-ZN95T4/glibc-2.31/intl/loadmsgcat.c:970:34
 #8 0x00007f39f7e0af36 (/lib/x86_64-linux-gnu/libc.so.6+0x36f36)
 #9 0x00007f39fdf877b8 SPIRV::LLVMToSPIRV::transUnaryInst(llvm::UnaryInstruction*, SPIRV::SPIRVBasicBlock*) (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVMSPIRVLib.so.11+0x23c7b8)
#10 0x00007f39fdf81e30 SPIRV::LLVMToSPIRV::transValueWithoutDecoration(llvm::Value*, SPIRV::SPIRVBasicBlock*, bool, SPIRV::LLVMToSPIRV::FuncTransMode) (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVMSPIRVLib.so.11+0x236e30)
#11 0x00007f39fdf82a65 SPIRV::LLVMToSPIRV::transValue(llvm::Value*, SPIRV::SPIRVBasicBlock*, bool, SPIRV::LLVMToSPIRV::FuncTransMode) (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVMSPIRVLib.so.11+0x237a65)
#12 0x00007f39fdf8852c SPIRV::LLVMToSPIRV::transFunction(llvm::Function*) (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVMSPIRVLib.so.11+0x23d52c)
#13 0x00007f39fdf8b209 SPIRV::LLVMToSPIRV::translate() (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVMSPIRVLib.so.11+0x240209)
#14 0x00007f39fdf8b3d8 SPIRV::LLVMToSPIRV::runOnModule(llvm::Module&) (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVMSPIRVLib.so.11+0x2403d8)
#15 0x00007f39f8ee8735 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVM-11.so+0xc4a735)
#16 0x00007f39fdf8ba70 llvm::writeSpirv(llvm::Module*, SPIRV::TranslatorOpts const&, std::ostream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/../lib/libLLVMSPIRVLib.so.11+0x240a70)
#17 0x0000564b4576df21 main (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/llvm-spirv+0xcf21)
#18 0x00007f39f7dfb0b3 __libc_start_main /build/glibc-ZN95T4/glibc-2.31/csu/../csu/libc-start.c:342:3
#19 0x0000564b4576e46d _start (/localdisk/work/akharche/miniconda3/envs/dppy_llvm11/bin/llvm-spirv+0xd46d)
```

**Comment by akharche** [comment]
Another problem is reproduced on O0 with llvm-spirv from oneAPI 2021.2 
InvalidModule: Invalid SPIR-V module: Casts from private/local/global address space are allowed only to generic

**Comment by PokhodenkoSA** [comment]
It is not problem in llvm-spirv. Should be closed by #364
